src/lib/oauth2.ts

```diff
@@ -27,6 +30,77 @@ export interface CookieOptions {
     partitioned?: boolean | undefined;
 }
 
+/** The base state query attribute for SSO */
+type SsoBaseState = {
+    /** Url to redirect too, after SSO has been performed */
+    redirectUrl: string;
+};
+
+/** The state query attribute for SSO */
+type SsoState = SsoBaseState &
+    (
+        | {
+              /** If this is a register request */
+              method: 'register';
+              /** Register requests have the name of the ioBroker user to register */
+              user: string;
+          }
+        | {
+              /** If this is a login request */
+              method: 'login';
+          }
+    );
+
+/** Query parameters in the SSO callback */
+interface SsoCallbackQuery {
+    /** Code to exchange for token */
+    code: string;
+    /** SsoState as parseable string */
+    state: string;
+}
+
+/** Response from Keycloak */
+interface OidcTokenResponse {
+    access_token: string;
+    refresh_token: string;
+    token_type: 'Bearer';
+    /** Used to retrieve the {@link JwtFullPayload} */
+    id_token: string;
+    'not-before-policy': number;
+    session_state: string;
+    scope: string;
+}
+
+/**
+ * Retrieved by decoding the JWT {@link OidcTokenResponse.id_token}
+ * The `sub` attribute is used to identify a user in `common.externalAuthentication.oidc.sub`
+ */
+interface JwtFullPayload extends Required<JwtPayload> {
+    auth_time: number;
+    typ: string;
+    azp: string;
+    sid: string;
+    at_hash: string;
+    acr: string;
+    email_verified: boolean;
+    name: string;
+    preferred_username: string;
+    given_name: string;
+    family_name: string;
+    email: string;
+}
+
+/** Keycloak ioBroker realm */
+const KEYCLOAK_ISSUER = 'https://keycloak.heusinger-it.duckdns.org/realms/iobroker-local';
```

Comment:
Of course after the PoC this should be a production instance hosted on public `iobroker.net` domain. And then other services could use the Keycloak instance too, like Cloud or Forum users. Inside of the Keycloak also other sign-ins could be configured (rn it's just GH or use normal email & pass), like Google, Apple to name the most important ones.